src/highdicom/seg/sop.py

```diff
@@ -944,3 +966,1458 @@ def _encode_pixels(self, planes: np.ndarray) -> bytes:
+                    'in this image.'
+                )
+        # Create the sub-matrix of the look up table that indexes
+        # by the dimension index values
+        dim_ind_cols = [
+            self._lut_dim_ind_cols[ptr]
+            for ptr in dimension_index_pointers
+        ]
+        lut = self._frame_lut[:, dim_ind_cols + [self._lut_seg_col]]
+
+        return np.unique(lut, axis=0).shape[0] == lut.shape[0]
+
+    def _check_indexing_with_source_frames(
+        self,
+        ignore_spatial_locations: bool = False
+    ) -> None:
+        """Check if indexing by source frames is possible.
+
+        Raise exceptions with useful messages otherwise.
+
+        Possible problems include:
+            * Spatial locations are not preserved.
+            * The dataset does not specify that spatial locations are preserved
+              and the user has not asserted that they are.
+            * At least one frame in the segmentation lists multiple
+              source frames.
+
+        Parameters
+        ----------
+        ignore_spatial_locations: bool
+            Allows the user to ignore whether spatial locations are preserved
+            in the frames.
+
+        """
+        # Checks that it is possible to index using source frames in this
+        # dataset
+        if self._locations_preserved is None:
+            if not ignore_spatial_locations:
+                raise RuntimeError(
+                    """
+                    Indexing via source frames is not permissible since this
+                    image does not specify that spatial locations are preserved
+                    in the course of deriving the segmentation from the source
+                    image. If you are confident that spatial locations are
+                    preserved, or do not require that spatial locations are
+                    preserved, you may override this behavior with the
+                    'ignore_spatial_locations' parameter.
+                    """
+                )
+        elif self._locations_preserved == SpatialLocationsPreservedValues.NO:
+            if not ignore_spatial_locations:
+                raise RuntimeError(
+                    """
+                    Indexing via source frames is not permissible since this
+                    image specifies that spatial locations are not preserved in
+                    the course of deriving the segmentation from the source
+                    image. If you do not require that spatial locations are
+                    preserved you may override this behavior with the
+                    'ignore_spatial_locations' parameter.
+                    """
+                )
+        if not self._single_source_frame_per_seg_frame:
+            raise RuntimeError(
+                'Indexing via source frames is not permissible since some '
+                'frames in the segmentation specify multiple source frames.'
+            )
+
+    def get_pixels_by_source_instance(
```

Comment:
I am wondering whether we should introduce two functions to handle the cases of single-frame and multi-frame source images separately.

```python
def get_pixels_by_reference(
    self,
    references: Sequence[Tuple[str, int]]
    ...
) -> np.ndarray
```

where `references` provides SOP Instance UID and Frame Number to uniquely identify each frame (see also [Image SOP Instance Reference Macro Attributes](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_10.3.html#table_10-3)). In case of a series of single-frame source images, the frame number shall always be `1` and in case of a multi-frame source image the SOP Instance UID shall be the same for each frame. We could also support `Sequence[Union[str, Tuple[str, int]]]` to allow omitting the Frame Number in case of a series of single-frame images.

Have you considered this approach?

Reply:
I actually started down that path and then split into two functions after a while because I think the single-method approach leads to quite a confusing function signature/API. In the case of a multiframe, unless something quite complex is going on that I don't think we need to support, all the frames with have the same UID leading to a lot of redundancy in the input. The two-method approach perhaps leads to a small amount of code redundancy but i think gives a cleaner interface and simpler implementation